GPII-3899: Initial implementation of global validation component.

the-t-in-rtf opened the pull request:
See [GPII-3899](https://issues.gpii.net/browse/GPII-3899) for details.

amb26 commented on src/js/common/hashString.js:
```diff
@@ -0,0 +1,63 @@
+     * @return {String} - A hash composed of the hashes for each sub-segment in the original string.
+     *
+     */
+    gpii.schema.hashString = function (originalString) {
```
Whilst you've worked hard to reduce the probability of a hash collision, I feel that in practice we simply can't afford any - at least, "essentially any". The following table explains that, for example, if we are just considering the space of strings shorter than 50k which just get a single 32-bit hash, after only 100,000 strings the chance of a collision becomes greater than 1 in 2. https://stackoverflow.com/questions/14210298/probability-of-collision-when-using-a-32-bit-hash
In any case, the "correct" data structure to be used (either an open or a closed hashtable https://stackoverflow.com/questions/9124331/meaning-of-open-hashing-and-closed-hashing ) would unavoidably store the actual key of the hash somewhere, in order to detect whether a collision has occurred.
In this case I recommend we simply store the key itself - either the JSON.stringify version of the schema or a lightly compacted version thereof, and thus eliminate all the complexity due to this hashing. We can revisit this in future if it becomes a performance worry - but bear in mind that we can't avoid storing the schema itself in memory, and this will at most double that storage cost.

the-t-in-rtf replied:
I'm happy to get rid of it and review if we find that it would be useful to have a shorter hash in the future.

amb26 commented on src/js/common/schemaValidatedModelComponent.js:
```diff
@@ -16,17 +16,24 @@ var fluid  = fluid  || {};
     fluid.defaults("gpii.schema.modelComponent", {
         gradeNames: ["gpii.schema.component", "fluid.modelComponent"],
+        members: {
+            schemaHash: false
```
Why not generate the schema hash in an expander in place?

the-t-in-rtf replied:
My goal was to make it easier to correlate a cached schema with its original component.  If we don't want to use more readable hashes,  I'm not sure there's much value in storing the stringified schema in the component and passing it around.

amb26 commented on src/js/common/validator.js:
```diff
@@ -148,21 +150,47 @@ var fluid  = fluid  || {};
-    gpii.schema.validator.validate = function (toValidate, gssSchema, ajvOptions) {
+    gpii.schema.validator.validate = function (that, gssSchema, toValidate, schemaHash) {
+        if (!schemaHash) {
+            schemaHash = gpii.schema.hashString(gpii.schema.stringify(gssSchema));
```
See above

the-t-in-rtf replied:
Yup, gone.

amb26 commented on src/js/common/validator.js:
```diff
@@ -572,4 +596,24 @@ var fluid  = fluid  || {};
+        gradeNames: ["fluid.component", "fluid.resolveRootSingle"],
+        singleRootType: "gpii.schema.validator",
+        members: {
+            validatorsByHash: {
```
Provide an operation to evict validators on request otherwise the user has no option but for them to grow without limit!

the-t-in-rtf replied:
Agreed, and done and tested.

the-t-in-rtf commented:
The tests failed because stringify takes longer on the windows VM.  I can relax or remove that check when I'm working on the hashing.

the-t-in-rtf commented:
The last build error was the same timing error in response to a later commit.  The upcoming build should include the timing change.

the-t-in-rtf commented:
@amb26, I've addressed the first round of comments.

amb26 commented on src/js/common/validator.js:
```diff
@@ -572,4 +592,28 @@ var fluid  = fluid  || {};
+                args:     ["{that}", "{arguments}.0", "{arguments}.1", "{arguments}.2"] // gssSchema, toValidate, schemaHash
+            },
+            clearCache: {
+                funcName: "fluid.set",
```
Whilst clearing the whole cache is probably a worthwhile operation to support, it is more likely that the user will want the more controllable option of evicting a the cache for a particular schema that they know they will not be using again. The operation should accept the gssSchema which is to be evicted, and call a "delete" on the matching cache slot.

the-t-in-rtf replied:
I went just a little further, as I would like to have a place to sketch more performant alternatives off to the side.  I added an invoker to remove a single schema from the cache, and also one to cache a single schema.  I also restored the ability to pass in a schema hash throughout.  This leaves the option to further optimise in derived grades without requiring us to do the work now.

amb26 commented on src/js/common/orderedStringify.js:
```diff
@@ -0,0 +1,59 @@
+    // TODO: pull this up into Infusion if we need to use this type of object hashing more broadly.
+    "use strict";
+
+    if (!fluid.identity) {
```
Funny way of checking for Infusion - probably cleanest to move this up to the outer level where we attempt to allocate the fluid global in the first place. I believe that is the form of pattern we use elsewhere.

the-t-in-rtf replied:
I standardised to the usual check for fluid itself and kept the `fluid.require` pattern for the non-fluid dependencies.

amb26 commented on src/js/common/validator.js:
```diff
@@ -148,21 +149,75 @@ var fluid  = fluid  || {};
+     * @return {Object} - The compiled validator created from the GSS schema.
+     *
+     */
+    gpii.schema.validator.cacheSchema = function (that, gssSchema, schemaHash) {
```
Nice addition - if, as you seem to be allowing in all of these functions, that the user may have computed the hash already, best to produce a generic function name (e.g. gpii.schema.hashSchema, which will currently just be an alias for gpii.schema.stringify) which they might have used to do so -

the-t-in-rtf replied:
Good idea, I added this in the validator file so that it could take advantage of the custom JSDoc object type.  I also updated the docs.

the-t-in-rtf commented:
The last build failed because grunt wasn't available for linting, I assume a transient provisioning failure.  Ok to test.

the-t-in-rtf commented:
Second failure, I dug into it, looks like the VM had trouble connecting to chocolatey to download the required packages.

```
==> windows10: [NuGet] An error occurred while loading packages from 'https://chocolatey.org/api/v2/': Unable to connect to the remote server
==> windows10: chromedriver not installed. The package was not found with the source(s) listed.
==> windows10:  Source(s): 'https://chocolatey.org/api/v2/'
```

The process aborted on that install, and missed the step to install grunt as well, which is why the linting checks failed.

the-t-in-rtf commented:
I see the same problem in [recent windows builds in CI](https://ci.gpii.net/job/windows-create-vm/719/console):

```
python2 not installed. The package was not found with the source(s) listed.
 Source(s): 'https://chocolatey.org/api/v2/'
```

the-t-in-rtf commented:
Ok to test again.

the-t-in-rtf commented:
Seems like we're past the networking issues that killed the previous two builds.  @amb26, this should be ready for further review.

the-t-in-rtf commented:
Thanks, @amb26.  I cut the release and updated the forward-facing version in master.  I also updated the companion pull against universal (will comment there shortly).
